### PR TITLE
virt_autotest: Enable NVME Hard Drive Support

### PR DESCRIPTION
**Pull request by guoxuguang**
Enable NVME Hard Drive Support

- Related ticket: https://progress.opensuse.org/issues/163502
- Verification run: 
SATA Hard Drive
[amd-zen3-gpu-sut1-1](https://openqa.suse.de/tests/15007506#) PASS
[openqaipmi5](https://openqa.suse.de/tests/15007521#) PASS
[gonzo](https://openqa.suse.de/tests/15008383#) PASS
[fozzie](https://openqa.suse.de/tests/15008384#) PASS
[quinn](https://openqa.suse.de/tests/15008380#) PASS
NVME Hard Drive
[bare-metal1](https://openqa.suse.de/tests/15007507#step/libvirt_extend_storage_lvm/179) PASS
[bare-metal2](https://openqa.suse.de/tests/15007508#step/libvirt_extend_storage_lvm/179) PASS

### CODE REVIEW

**Comment by RoyCai7** [comment]
NVME Hard Drive
[gi-guest_developing-on-host_developing-kvm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/19794)

the VR link is not right.

File: lib/virt_autotest/virtual_storage_utils.pm [hunk 1]
```diff
@@ -90,6 +90,8 @@ sub destroy_virt_storage_pool {
 sub delete_physical_volume {
     my $lvm_disk_name = shift;
     my $timeout = 180;
+    # Enable NVME Hard Drive Support
+    $lvm_disk_name = ($lvm_disk_name =~ "nvme") ? "${lvm_disk_name}p" : $lvm_disk_name;
```
Comment (waynechen55) [inline lib/virt_autotest/virtual_storage_utils.pm:94]: Does this apply to all nvme cases ?

File: tests/virt_autotest/libvirt_extend_storage_lvm.pm [hunk 2]
```diff
@@ -132,6 +132,8 @@ sub create_volume_group {
     # Create new disk partition for LVM volumes
     record_info "Create new disk partition for LVM volumes";
     assert_script_run 'echo -e "g\nn\n\n\n+20G\nt\n8e\np\nw" | fdisk ' . $lvm_disk_name;
+    # Enable NVME Hard Drive Support
+    $lvm_disk_name = ($lvm_disk_name =~ "nvme") ? "${lvm_disk_name}p" : $lvm_disk_name;
```
Comment (waynechen55) [inline tests/virt_autotest/libvirt_extend_storage_lvm.pm:136]: Does this apply to all nvme cases ?

**Comment by alice-suse** [comment]
Hi @guoxuguang ,
Would you please explain where the problem is and how you fix it? I am a little bit confused now. The device used in test does not seem consistent. See below execution log from your verification job.
- on top of the screen, nvme1n11 is used in `pvremove -f -y`
- in the middle, there is partition error when creating nvme1n1p1
- there is also references to nvme1n11 in autoinst.log, eg  `pvdisplay /dev/nvme1n11`

image

**Comment by guoxuguang** [comment]
> NVME Hard Drive [gi-guest_developing-on-host_developing-kvm](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/19794)
> 
> the VR link is not right.

Add the VR link for NVME Hard Driver
[gi-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/14994901) 

Sorry for my mistake due to cannot be accessed successfully yesterday.

File: lib/virt_autotest/virtual_storage_utils.pm (hunk 1 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_storage_utils.pm:94]: Yes, this changed would be applied to our all testing NVME Hard Drive. But, Still need more VR to validate.

File: tests/virt_autotest/libvirt_extend_storage_lvm.pm (hunk 2 above)
Comment (guoxuguang) [reply tests/virt_autotest/libvirt_extend_storage_lvm.pm:136]: As the same, yes, this changed would be applied to our all testing NVME Hard Drive. But, Still need more VR to validate.

**Comment by guoxuguang** [comment]
> Hi @guoxuguang , Would you please explain where the problem is and how you fix it? I am a little bit confused now. The device used in test does not seem consistent. See below execution log from your verification job.
> 
>     * on top of the screen, nvme1n11 is used in `pvremove -f -y`
> 
>     * in the middle, there is partition error when creating nvme1n1p1
> 
>     * there is also references to nvme1n11 in autoinst.log, eg  `pvdisplay /dev/nvme1n11`

The root cause is still that after the NVME hard drive creates a partition, the partition name naming rules of the device are completely different between NVME and SATA hard drives. 

Situation: 
For NVME hard drives - `nvme1n1`,  the partition was named `/dev/nvme1n1p1` after creating a new partition. 
However,  for SATA hard drives - `sdb,`  the partition was named `/dev/sdb1` after creating a new partition. 
This naming convention is completely different between NVME and SATA hard drives. 

Solution:
Refer to the code logic from `tests/virt_autotest/libvirt_extend_storage_lvm.pm`, 
we'd better to use with `${lvm_disk_name}p` for our all NVME hard drives.

**Comment by guoxuguang** [comment]
> > > * on top of the screen, nvme1n11 is used in `pvremove -f -y`
> 
> Is nvme1n11 expected to be used? What is it?

In fact, for **NVME** devices,  there will be no partitions called `nvme1n11`. There should be called **nvme1n1p1** - `/dev/nvme1n1p1`:
```
# fdisk /dev/nvme1n1

Welcome to fdisk (util-linux 2.39.3).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.


Command (m for help): p 
Disk /dev/nvme1n1: 476.94 GiB, 512110190592 bytes, 1000215216 sectors
Disk model: SAMSUNG MZVL2512HCJQ-00B00              
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: FC18DC0E-32D1-43D9-A879-C2B097BD52AE

Device         Start      End  Sectors Size Type
/dev/nvme1n1p1  2048 41945087 41943040  20G Linux filesystem
```
> 
> > > * in the middle, there is partition error when creating nvme1n1p1
> 
> Seems different partition options needed for nvme to avoid the errors?

Refer to the following information from fdisk output:
`The device contains 'gpt' signature and it will be removed by a write command. See fdisk(8) man page and --wipe option for more details.`
This indicates that the disk we are attempting to modify has a GPT (GUID Partition Table) signature. When we use the fdisk tool to make changes, the GPT signature will be erased. Just a friendly reminder about the fdisk tool.

Situation:
This serves as a useful reminder that utilizing dd's will not fully erase the disk information on NVME drives. 
Essentially, even after using the dd method, the GPT (GUID Partition Table) signature persists on NVME disks.

Solution:
Using sgdisk - `sgdisk --zap-all ` to Wipe GPT(GUID Partition Table) Data(i.e. GPT  signature) for NVME drives: 
```
bare-metal2:~ # echo -e "g\nn\n\n\n+20G\nt\n8e\np\nw" | fdisk /dev/nvme1n1

Welcome to fdisk (util-linux 2.39.3).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

The device contains 'gpt' signature and it will be removed by a write command. See fdisk(8) man page and --wipe option for more details.

Device does not contain a recognized partition table.
Created a new DOS (MBR) disklabel with disk identifier 0x774564eb.

Command (m for help): Created a new GPT disklabel (GUID: 6C87E4E4-E54A-485D-8C1C-1DE93858959F).
The device contains 'gpt' signature and it will be removed by a write command. See fdisk(8) man page and --wipe option for more details.

Command (m for help): Partition number (1-128, default 1): First sector (2048-1000215182, default 2048): Last sector, +/-sectors or +/-size{K,M,G,T,P} (2048-1000215182, default 1000214527): 
Created a new partition 1 of type 'Linux filesystem' and of size 20 GiB.

Command (m for help): Selected partition 1
Partition type or alias (type L to list all): Type of partition 1 is unchanged: Linux filesystem.

Command (m for help): Disk /dev/nvme1n1: 476.94 GiB, 512110190592 bytes, 1000215216 sectors
Disk model: SAMSUNG MZVL2512HCJQ-00B00              
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 6C87E4E4-E54A-485D-8C1C-1DE93858959F

Device         Start      End  Sectors Size Type
/dev/nvme1n1p1  2048 41945087 41943040  20G Linux filesystem

Command (m for help): The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.

bare-metal2:~ # sgdisk --zap-all /dev/nvme1n1
GPT data structures destroyed! You may now partition the disk using fdisk or
other utilities.
bare-metal2:~ # echo -e "g\nn\n\n\n+20G\nt\n8e\np\nw" | fdisk /dev/nvme1n1

Welcome to fdisk (util-linux 2.39.3).
Changes will remain in memory only, until you decide to write them.
Be careful before using the write command.

Device does not contain a recognized partition table.
Created a new DOS (MBR) disklabel with disk identifier 0x809300b4.

Command (m for help): Created a new GPT disklabel (GUID: D355312E-E769-4E36-93BB-AD4DD3B336B3).

Command (m for help): Partition number (1-128, default 1): First sector (2048-1000215182, default 2048): Last sector, +/-sectors or +/-size{K,M,G,T,P} (2048-1000215182, default 1000214527): 
Created a new partition 1 of type 'Linux filesystem' and of size 20 GiB.

Command (m for help): Selected partition 1
Partition type or alias (type L to list all): Type of partition 1 is unchanged: Linux filesystem.

Command (m for help): Disk /dev/nvme1n1: 476.94 GiB, 512110190592 bytes, 1000215216 sectors
Disk model: SAMSUNG MZVL2512HCJQ-00B00              
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: D355312E-E769-4E36-93BB-AD4DD3B336B3

Device         Start      End  Sectors Size Type
/dev/nvme1n1p1  2048 41945087 41943040  20G Linux filesystem

Command (m for help): The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.
```

> > > * there is also references to nvme1n11 in autoinst.log, eg  `pvdisplay /dev/nvme1n11`
> 
> same here as the first one...
There should be used with **nvme1n1p1** instead. 
```
bare-metal2:~ # pvdisplay /dev/nvme1n1p1
  Failed to find physical volume "/dev/nvme1n1p1".
```

**Comment by guoxuguang** [comment]
Add the latest validation run as below:
SATA Hard Drives
[gi-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/15000946#)
NVME Hard Drives
[bare-metal1](https://openqa.suse.de/tests/15000991#)
[bare-metal2](https://openqa.suse.de/tests/15000992#)

**Comment by guoxuguang** [comment]
Add the latest validation run as below:
SATA Hard Drive
[amd-zen3-gpu-sut1-1](https://openqa.suse.de/tests/15007506#) PASS
[openqaipmi5](https://openqa.suse.de/tests/15007521#) PASS
NVME Hard Drive
[bare-metal1](https://openqa.suse.de/tests/15007507#step/libvirt_extend_storage_lvm/179) PASS
[bare-metal2](https://openqa.suse.de/tests/15007508#step/libvirt_extend_storage_lvm/179) PASS

If there were not any new comments, this PR is ready to merge now.

**Comment by alice-suse** [comment]
Thank you, looks good now

File: tests/virt_autotest/libvirt_extend_storage_lvm.pm [hunk 3]
```diff
@@ -102,7 +104,13 @@ sub get_hard_disk_size {
 # Wipe Hard Disk Clean via dd
 sub wipe_hard_disk {
     my $hard_disk_name = shift;
+    my $timeout = 180;
+    record_info("Wipe Hard Drive $hard_disk_name forcibly");
     assert_script_run("dd if=/dev/zero of=$hard_disk_name count=1M", timeout => 1500, fail_message => "Failed to wipe hard disk clean on $hard_disk_name");
+    record_info("Remove GPT signature $hard_disk_name forcibly");
+    script_run("sgdisk --zap-all $hard_disk_name", timeout => $timeout);
```
Comment (alice-suse) [inline tests/virt_autotest/libvirt_extend_storage_lvm.pm:111]: Based on code logic, it seems this cmd is necessary and shall be successful . Then assert_script_run is better?

**Comment by Julie-CAO** [comment]
LGTM

File: tests/virt_autotest/libvirt_extend_storage_lvm.pm (hunk 3 above)
Comment (alice-suse) [reply tests/virt_autotest/libvirt_extend_storage_lvm.pm:111]: This is not needed any more. I just realize the default behavior of script_run is also die after timeout.

**Review by alice-suse — APPROVED** [review]
LGTM

**Pull request merged**